src/GraphQL/Validation/Rules/UniqueFragmentNames.cs

```diff
@@ -19,22 +19,21 @@ public class UniqueFragmentNames : IValidationRule
 
         /// <inheritdoc/>
         /// <exception cref="UniqueFragmentNamesError"/>
-        public Task<INodeVisitor> ValidateAsync(ValidationContext context)
-        {
-            var knownFragments = new Dictionary<string, FragmentDefinition>();
+        public Task<INodeVisitor> ValidateAsync(ValidationContext context) => context.Document.Fragments.Count > 1 ? _nodeVisitor : null;
 
-            return new MatchingNodeVisitor<FragmentDefinition>(fragmentDefinition =>
+        private static readonly Task<INodeVisitor> _nodeVisitor = new MatchingNodeVisitor<FragmentDefinition>((fragmentDefinition, context) =>
+            {
+                var knownFragments = context.TypeInfo.UniqueFragmentNames_KnownFragments ??= new Dictionary<string, FragmentDefinition>();
+
+                var fragmentName = fragmentDefinition.Name;
+                if (knownFragments.ContainsKey(fragmentName)) // .NET 2.2+ has TryAdd
```

Comment:
There's a couple places where TryAdd would be faster, but not available with .Net Standard 2.0